[libclang/python] Add LIBCLANG_LIBRARY_PATH and LIBCLANG_LIBRARY_FILE

TApplencourt opened the pull request:
Close #167421 

This PR adds the environment variables ``LIBCLANG_LIBRARY_PATH`` and ``LIBCLANG_LIBRARY_FILE``, which allow users to specify the directory path and the exact library file that should be used to locate libclang.

TApplencourt committed:
Add LIBCLANG_LIBRARY_PATH and LIBCLANG_LIBRARY_FILE

TApplencourt commented:
For more rational on the change see the issue https://github.com/llvm/llvm-project/issues/167421.

The PR is currently in the "minimal" stage. I can:
- In the test remove the old support for `CLANG_LIBRARY_PATH`, and add in the release note that people should migrate to `LIBCLANG_LIBRARY_PATH` (will remove 3 lines per tests file so simplifying the code base)
- Add tests. The simple one will be to do a negative testing (aka setting `LIBCLANG_LIBRARY_PATH` to `/dev/null` and verify that it crash). Other test seem harder (1/ verify that cindex if working 2/ find the `libclang.so` that he used, 3/ use this path to set `LIBCLANG_LIBRARY_PATH`, retry)

Curious what do people thinks (previous `CLANG_LIBRARY_PATH` didn't get any test, but it was used only the tests)

DeinAlptraum reviewed (commented):
> add in the release note that people should migrate to LIBCLANG_LIBRARY_PATH

I don't think that's necessary since this is only used by the tests and they are not part of the interface we offer. 
Where this should be adapted however is the `clang/bindings/python/README.md`

> Add tests.

Yes please. You can create an additional test file for this, where you set the environment variable from the code, and _then_ import `cindex`. Then this shouldn't be an issue. I think...
All in favor of removing the old setting `CLANG_LIBRARY_PATH` at the top of all the test files. Note that the current cmake build flow requires this, so you will need to change the variable name in `clang/bindings/python/tests/CMakeLists.txt`



Additionally, in `get_cindex_library`, an error message is returned that refers to the methods to set the library file. This should be adapted to also mention the environment variables.

TApplencourt committed:
Move too LIBCLANG_LIBRARY_PATH

TApplencourt commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -170,6 +170,9 @@ Clang Python Bindings Potentially Breaking Changes
 - Remove ``AccessSpecifier.NONE`` kind. No libclang interfaces ever returned this kind.
+- Added the environment variables ``LIBCLANG_LIBRARY_PATH`` and ``LIBCLANG_LIBRARY_FILE``,
+  which allow users to specify the directory path and the exact library file that
+  should be used to locate libclang.
```
Far better. Thanks!

TApplencourt committed:
Update clang/docs/ReleaseNotes.rst

Co-authored-by: Jannick Kremer <jannick.kremer@mailbox.org>

TApplencourt committed:
update error message

TApplencourt commented:
Ok, so:
- Added new tests: `clang/bindings/python/tests/cindex/test_environment_variable.py`
- Removed the old environments notions (update the README and the CINDEX error message)
- Updated the error message

Sorry I failed my `git amend` so the commit history is ugly, will make the review harder. 

---

Erf, I run `black` I guess `darker` is not `black` : ( Will fix!

TApplencourt committed:
Apply darker sugestion

TApplencourt committed:
Merge branch 'main' into env_cindex

TApplencourt committed:
Fix negative tests...

TApplencourt committed:
One more set of fix

DeinAlptraum commented on clang/bindings/python/tests/cindex/test_environment_variable.py:
```diff
@@ -0,0 +1,63 @@
+import os
+
+
+def reset_import_and_get_frech_config():
```
```suggestion
def reset_import_and_get_fresh_config():
```

DeinAlptraum reviewed (commented):
Some minor cleanup, otherwise LGTM!

Endilll reviewed (commented):
I'm not a fan of the existing terminology in this part of our codebase, where "path" should be understood as "search path" or "containing directory", while "file" and "filename" are "path to the file" (probably a full path, but that is yet to be confirmed). I left couple of suggestions that should improve the clarity of the error message and the release note.

That's probably out of scope here, but I'd like to see `Config.library_path`, `Config.library_file`, `Config.get_filename` to be renamed to something less confusing. As for `LIBCLANG_LIBRARY_PATH`, it can be argued that it follows a well-established example of `LD_LIBRARY_PATH`, so I can live with this.

TApplencourt committed:
Update clang/bindings/python/clang/cindex.py

Co-authored-by: Vlad Serebrennikov <serebrennikov.vladislav@gmail.com>

TApplencourt committed:
Update clang/bindings/python/clang/cindex.py

Co-authored-by: Vlad Serebrennikov <serebrennikov.vladislav@gmail.com>

TApplencourt committed:
Update clang/docs/ReleaseNotes.rst

Co-authored-by: Vlad Serebrennikov <serebrennikov.vladislav@gmail.com>

TApplencourt committed:
Fix typos

TApplencourt committed:
Fix typos in reset_import_and_get_frech_config

TApplencourt committed:
Fix unused import

DeinAlptraum reviewed (approved):
LGTM!

@Endilll if you have no further comments, I would merge this
